readthedocs/proxito/views.py

```diff
@@ -0,0 +1,247 @@
+                subproject = get_object_or_404(Project, slug=subproject_slug)
+        return view_func(request, subproject=subproject, *args, **kwargs)
+
+    return inner_view
+
+
+def map_project_slug(view_func):
+    """
+    A decorator that maps a ``project_slug`` URL param into a Project.
+
+    :raises: Http404 if the Project doesn't exist
+
+    .. warning:: Does not take into account any kind of privacy settings.
+    """
+
+    @wraps(view_func)
+    def inner_view(  # noqa
+            request, project=None, project_slug=None, *args, **kwargs
+    ):
+        if project is None:
+            # Get a slug from the request if it can't be found in the URL
+            if not project_slug:
+                project_slug = request.host_project_slug
+                log.debug(
+                    'Inserting project slug from request slug=[%s]',
+                    project_slug
+                )
+            try:
+                project = Project.objects.get(slug=project_slug)
+            except Project.DoesNotExist:
+                raise Http404('Project does not exist.')
+        return view_func(request, project=project, *args, **kwargs)
+
+    return inner_view
+
+
+@map_project_slug
+@map_subproject_slug
+def redirect_page_with_filename(request, project, subproject, filename):  # pylint: disable=unused-argument  # noqa
+    """Redirect /page/file.html to /<default-lang>/<default-version>/file.html."""
+
+    urlparse_result = urlparse(request.get_full_path())
+    return HttpResponseRedirect(
+        resolve(
+            subproject or project,
+            filename=filename,
+            query_params=urlparse_result.query,
+        )
+    )
+
+
+@map_project_slug
+@map_subproject_slug
+def redirect_project_slug(request, project, subproject):  # pylint: disable=unused-argument
+    """Handle / -> /en/latest/ directs on subdomains."""
+    urlparse_result = urlparse(request.get_full_path())
+
+    return HttpResponseRedirect(
+        resolve(
+            subproject or project,
+            query_params=urlparse_result.query,
+        ),
+    )
+
+
+@map_project_slug
+@map_subproject_slug
+def serve_docs(
+        request,
+        project,
+        subproject,
+        lang_slug=None,
+        version_slug=None,
+        filename='',
+):
+    """Take the incoming parsed URL's and figure out what file to serve."""
+
+    log.debug(
+        'project=%s, subproject=%s, lang_slug=%s, version_slug=%s, filename=%s',
+        project, subproject, lang_slug, version_slug, filename
+    )
+
+    # Take the most relevant project so far
+    current_project = subproject or project
+
+    # Handle a / redirect when we aren't a single version
+    if all([lang_slug is None, version_slug is None, filename == '',
+            not current_project.single_version]):
+        log.info('Proxito redirect: slug=%s', current_project.slug)
+        return redirect_project_slug(
+            request, project=current_project, subproject=None
+        )
```

Comment:
I think I found a change in the behavior of the `/` redirect.

Currently in production, if we hit the `.readthedocs.io` URL (https://requests.readthedocs.io/) of a project with a Custom Domain defined, our NGINX will serve the documentation under that URL.

On the other hand, El Proxito is redirecting me to the Custom Domain project URL.

The new behavior is what I like the most (and I think @davidfischer as well) but I just wanted to note that there is a change here that we may need to take into account.

Reply:
Yea, I'm mixed about what the proper behavior is. I think the root redirects are fine to automatically use the custom domain, but I'm guessing we'll actually want this to be configurable in the future. Sometimes I actually want to be able to go to the `readthedocs.io` domain even when there's a custom domain.